Allow concurrent build with RPC server

gridbugs opened the pull request:
Allow the `dune build` command to be run concurrently with an RPC server (in either passive or eager watch mode). This works by having the second instance of dune send RPC build command to the RPC server, and the RPC server then sends diagnostic messages back to the first instance of dune.

Currently this is only implemented for the `dune build` command, and promoting changes into source files is not supported. There are also some subtle differences between the output of `dune build` is performing the build itself and when it's acting as an RPC client, such as file paths being absolute rather than relative and the underline under code errors losing its color.

gridbugs committed:
Allow concurrent build with RPC server

Allow the `dune build` command to be run concurrently with an RPC
server (in either passive or eager watch mode). This works by having the
second instance of dune send RPC build command to the RPC server, and
the RPC server then sends diagnostic messages back to the first instance
of dune.

Currently this is only implemented for the `dune build` command, and
promoting changes into source files is not supported. There are also
some subtle differences between the output of `dune build` is performing
the build itself and when it's acting as an RPC client, such as file
paths being absolute rather than relative and the underline under code
errors losing its color.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

Leonidas-from-XIV reviewed (commented):
I think it's a very nice feature. I know it has some limitations (e.g. can't build in a different profile) but I was glad to see that the target can be specified and I think that covers 80% of all usecases.

It would be nice to have tests that show this mechanism working.

gridbugs commented on bin/build_cmd.ml:
```diff
@@ -138,6 +138,26 @@ let run_build_command ~(common : Common.t) ~config ~request =
+    Printf.printf
+      "Error: %s\n%!"
+      (Dyn.to_string (Dune_rpc_private.Response.Error.to_dyn error))
+  | Ok Success -> print_endline "Success"
```
I've changed it to use `Dune_console` for now, but I don't feel strongly about changing it to not output anything on success either.

gridbugs commented on bin/build_cmd.ml:
```diff
@@ -164,10 +184,14 @@ let build =
-    run_build_command ~common ~config ~request
+    match Dune_rpc_client.Where.get () with
+    | Some where ->
+      Scheduler.go_without_rpc_server ~common ~config (fun () -> build_rpc where targets)
```
Renamed to `build_via_rpc_server` and added some clarifying comments.

gridbugs commented on bin/build_cmd.ml:
```diff
@@ -164,10 +184,14 @@ let build =
-      Target.interpret_targets (Common.root common) config setup targets
-    in
-    run_build_command ~common ~config ~request
+    match Dune_rpc_client.Where.get () with
```
Yep. Clarified with a comment.

gridbugs commented on bin/rpc/build.ml:
```diff
@@ -29,50 +29,62 @@ let establish_connection_or_raise ~wait once =
   establish_connection_or_raise ~wait once
 ;;
 
+let build_sexp_string_targets ~wait ~targets where =
```
It initializes an rpc client, connects to the server at `where`, and then sends a request to build the `targets` which are specified as a list of sexp-encoded strings passed as arguments to the `dune rpc build` command.

gridbugs commented on bin/rpc/build.ml:
```diff
@@ -29,50 +29,62 @@ let establish_connection_or_raise ~wait once =
+  let targets =
+    List.map targets ~f:(fun target ->
+      let sexp = Dune_lang.Dep_conf.encode target in
+      Dune_lang.to_string sexp)
```
By "build" I mean compiled. This is where we encode targets as sexp strings but the targets are built by the rpc server in response to the build request sent by `build_sexp_string_targets`. I'll add a comment to clarify.

gridbugs commented on bin/rpc/build.ml:
```diff
@@ -29,50 +29,62 @@ let establish_connection_or_raise ~wait once =
   Rpc_common.client_term builder
   @@ fun _common ->
+  let where =
+    match Dune_rpc_impl.Where.get () with
```
Do you mean once here and once in build_cmd.ml? This file implements `dune rpc build` while build_cmd.ml implements `dune build`, so it won't be called twice in a single invocation of dune.

gridbugs commented on bin/rpc/build.ml:
```diff
@@ -29,50 +29,62 @@ let establish_connection_or_raise ~wait once =
+      "Error: %s\n%!"
+      (Dyn.to_string (Dune_rpc_private.Response.Error.to_dyn error))
+  | Ok Success -> print_endline "Success"
+  | Ok (Failure _) -> print_endline "Failure"
```
Maybe but I'm not so sure. This code was just refactored in this PR, not added. `dune rpc build` is not a very user-friendly command (e.g. it needs to be passed sexps on the command-line) so my assumption is that it's intended to be called from scripts. As such it's not unreasonable to expect that scripts would expect the status of the build to be printed to stdout for ease of parsing.

gridbugs committed:
Rename Scheduler.go to go_with_rpc_server

This function evaluates a fiber concurrently with an RPC server, so
renaming to clarify this fact.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs commented on bin/rpc/build.mli:
```diff
@@ -1,2 +1,15 @@
+(** Sends a command to an RPC server to build the specified targets and wait
+    for the build to complete or fail. *)
+val build
+  :  wait:bool
```
For the connection. I've added that info to the doc comment.

gridbugs commented on src/dune_rpc_impl/decl.ml:
```diff
@@ -1,20 +1,47 @@
+    let open Conv in
+    let from { main; related } = main, related in
+    let to_ (main, related) = make ~main ~related in
+    let main = field "main" (required User_message.sexp_without_annots) in
```
Clarified with a doc comment on the `sexp_without_annots` function.

gridbugs committed:
Document some [Conv] parser combinators

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs commented:
> It would be nice to have tests that show this mechanism working.

Added a test.

gridbugs commented:
Looks like this change has made some of our rpc tests flaky. Investigating...

gridbugs commented:
The problem was that I moved `Dune_rpc_impl.Where.get ()` outside of the retry loop so waiting for the server to start wasn't working properly. Should be fixed now.

gridbugs commented:
Ah looks like something else isn't working now.

gridbugs commented:
There's a non-deterministic crash with this error:
```
+|  Internal error, please report upstream including the contents of _build/log.
+|  Description:
+|    ("Memo error handler raised an exception",
+|     { exns =
+|         [ { exn =
+|               "File \"src/dune_engine/build_system.ml\", line 67, characters 11-17: Assertion failed"
+|           ; backtrace =
+|               "Raised at Dune_engine__Build_system.State.update_build_progress_exn in file \"src/dune_engine/build_system.ml\", line 67, characters 11-23\n\
+|                Called from Dune_engine__Build_system.State.add_errors in file \"src/dune_engine/build_system.ml\", lines 86-87, characters 6-73\n\
+|                Called from Dune_engine__Build_system.report_early_exn in file \"src/dune_engine/build_system.ml\", line 1110, characters 14-37\n\
+|                Called from Fiber__Scheduler.exec in file \"vendor/fiber/src/scheduler.ml\", line 76, characters 8-11\n\
+|                "
+|           }
+|         ]
+|     })
+|  Raised at Stdune__Code_error.raise in file
+|    "otherlibs/stdune/src/code_error.ml", line 10, characters 30-62
+|  Called from Fiber__Core.map.(fun) in file "vendor/fiber/src/core.ml", line
+|    68, characters 31-36
+|  Called from Fiber__Scheduler.exec in file "vendor/fiber/src/scheduler.ml",
+|    line 76, characters 8-11
+|
+|    I must not crash.  Uncertainty is the mind-killer. Exceptions are the
+|  little-death that brings total obliteration.  I will fully express my cases.
+|  Execution will pass over me and through me.  And when it has gone past, I
+|  will unwind the stack along its path.  Where the cases are handled there will
+|  be nothing.  Only I will remain.
```

The loc it points to is:
```ocaml
  let update_build_progress_exn ~f =
    let current = Svar.read t in
    match current with
    | Building current -> Svar.write t @@ Building (f current)
    | _ -> assert false   (* <------ *)
  ;;
```

I wonder if this is a race between a rebuild by the eager watch server in response to a file change and a build due to an rpc build request. Will investigate more tomorrow.

gridbugs committed:
Replace an [assert false] with a [Code_error]

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs commented:
> I wonder if this is a race between a rebuild by the eager watch server in response to a file change and a build due to an rpc build request.

This turned out to be the issue. I've fixed this but now it looks like it's getting stuck running tests. I can't reproduce locally but I'll try setting up a VM with a single core in case that makes the difference.

gridbugs commented:
> looks like it's getting stuck running tests

This issue seems to have gone away on its own? Maybe I needed to be more patient...
